src/index.ts

```diff
@@ -170,18 +179,40 @@ async function collectMappingOptions(
   }
 }
 
-async function apply(organizeOptions: OrganizeOptions) {
+export type ProgressCallback = (message: TWorkerMessage) => void
+
+let progressCallback: ProgressCallback | undefined
+
+export async function apply(
+  options: OrganizeOptions,
+  onProgress?: ProgressCallback,
+): Promise<TMapResults[]> {
+  progressCallback = onProgress
+
+  // Initialize workers
   const fileListWorker = initializeFileListWorker()
   initializeMappingWorkers()
-  // Set global mappingWorkerOptions
-  mappingWorkerOptions = (await collectMappingOptions(
-    organizeOptions,
-  )) as TMappingWorkerOptions
+
+  // Get mapping options
+  mappingWorkerOptions = await collectMappingOptions(options)
+
+  // Start scanning
   fileListWorker.postMessage({
     request: 'scan',
-    directoryHandle: organizeOptions.inputDirectory,
+    directoryHandle: options.inputDirectory,
   })
-  dispatchMappingJobs()
-}
 
-export { apply }
+  // Wait for completion
+  return new Promise((resolve) => {
```

Comment:
I don't particularly like that there is a non-DRY state check here.
Could we not align this state-check with the identical other check in the code?
E.g. can't you just call progressCallback({ response: 'finished', .. }) further above at console.log('finished')
instead of returning a promise?